### DISCUSSION TITLE
shapely.errors.TopologicalError

### DISCUSSION THREAD

**Discussion by w1125** [open-mmlab/mmocr#1603] [General]
训练数据集TotalText时出现错误shapely.errors.TopologicalError: The operation 'GEOSIntersection_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x7f4a807ef880>
数据集和标注文件来自官方文档提供的项目下载，json文件来自项目代码生成

**Comment by gaotongxiao** [comment]
We have `FixInvalidPolygon` transform to fix these invalid polygons. Please refer to an example config for usage:

https://github.com/open-mmlab/mmocr/blob/dev-1.x/configs/textdet/fcenet/fcenet_resnet50_fpn_1500e_totaltext.py

**Reply by w1125** [reply]
非常抱歉，我尝试在文件中加入了FixInvalidPolygon用法，仍然会报相同的错误。请问还有其他方法嘛或者是我没有正确使用。
在这个文件中添加的用法
![image](https://user-images.githubusercontent.com/101447162/216751963-413d6680-c809-43b2-8a36-c776c09ca50c.png)
代码如图所示：
![image](https://user-images.githubusercontent.com/101447162/216752000-1b45f60d-458d-4bb3-9364-24d755830b90.png)

**Reply by gaotongxiao** [reply]
`FixInvalidPolygon` in MMOCR 1.0 (`1.x` or `dev-1.x` branch) has been improved to handle more corner cases. MMOCR 0.6.3 (the current `main` branch) is no longer active for the upcoming updates.

If you still have to stick to MMOCR 0.6, you can try to migrate the [new implementation](
https://github.com/open-mmlab/mmocr/blob/c870046a4a194a60fa626dfd3c650143ff44887a/mmocr/datasets/transforms/ocr_transforms.py#L649) back to it, but note that the transform robustness in MMOCR 0.6 is far less than MMOCR 1.0, and therefore you might have to insert `FixInvalidPolygon` in between every transform in the pipeline.
[](url)

**Comment by w1125** [comment]
尝试在MMOCR 0.6 dbnetpp工程里加入FixInvalidPolygon，在mmocr/datasets/pipeline里注册FixInvalidPolygon，并在init文件里import
@PIPELINES.register_module()
class FixInvalidPolygon(BaseTransform)：
但是报错KeyError: 'UniformConcatDataset: "IcdarDataset: \'FixInvalidPolygon is not in the pipeline registry\'"'  表示注册并未成功
希望能给我一些建议解决这个问题，感谢您的帮助



---- Replied Message ----
| From | Tong ***@***.***> |
| Date | 02/05/2023 10:05 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [open-mmlab/mmocr] shapely.errors.TopologicalError (Discussion #1603) |

FixInvalidPolygon in MMOCR 1.0 (1.x or dev-1.x branch) has been improved to handle more corner cases. MMOCR 0.6.3 (the current main branch) is no longer active for the upcoming updates.

If you still have to stick to MMOCR 0.6, you can try to migrate the new implementation back to it, but note that the transform robustness in MMOCR 0.6 is far less than MMOCR 1.0, and therefore you might have to insert FixInvalidPolygon in between every transform in the pipeline.

**Reply by gaotongxiao** [reply]
Your changes seem not effective on code. Did you install `mmocr` in developer's mode [(step 2, case A)][(https://mmocr.readthedocs.io/en/latest/install.html#best-practices)? If you installed `mmocr` as a package (i.e. `pip install mmocr`), then your local changes to the code would not take affect.

If you don't want to re-install it in developer's mode, try starting your commands with `PYTHONPATH=.`, which should guide Python to use your MMOCR copy first. For example:
```bash
PYTHONPATH=. python tools/train.py configs/....
```
